### ISSUE TITLE
Explore shared library approach

### ISSUE THREAD

**Issue by petethepig** [pyroscope-io/pyroscope#216]
Some of the profilers we depend on use `process_vm_readv`, and it's disabled in some environments. We could maybe get around this if we added ability to integrate profilers as shared libraries. Then maybe the profiler could read its own memory. Then you'd be able to integrate by just adding a dependency to your code, something like

```
pip install pyroscope
```

TODO: add more details here

**Comment by AdrK** [comment]
I have been trying to play around with the process capabilities. Using `prctl()` you can change the capabilities of the process. I have written a piece of code in which I have used different ways to modify the capabilities of current process to allow for ptrace'ing it:
```
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <sys/ptrace.h>
#include <sys/prctl.h>

int main(int argc, char* argv[])
{
	//ptrace(PTRACE_TRACEME, 0, 0);
	prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY , 0, 0, 0);
	struct timespec ts = {.tv_sec = 0, .tv_nsec = 1e8};
	for(int i = 0;; i++)
	{
		printf("i = %d\n", i);
		nanosleep(&ts, NULL);
	}
}
```

That didn't worked :wink: So I figured that the capabilities are not an issue, especially because this:
```
root@c9ce6049f22e:/workdir# cat /proc/sys/kernel/yama/ptrace_scope
0
```
Shows that ptrace'ing capabilities should be enabled by default.

I was looking for a solution, and this doc nailed the topic:
`https://jvns.ca/blog/2020/04/29/why-strace-doesnt-work-in-docker/`

TL;DR is:
Docker is using seccomp to allow and dissallow certain syscalls and `--cap-add` in Docker is not only just setting the capabilities, but also modifies the `seccomp` settings to allow for `process_vm_readv & writev`

So at the end, seems that there is no other option than, either use `--cap-add` or `--security-opt seccomp=unconfined` or in any other way change the seccomp settings

**Comment by AdrK** [comment]
Since this commit ( Docker 19.03 ):
https://github.com/moby/moby/commit/1124543ca8071074a537a15db251af46a5189907
ptrace should work by default.

I will check it

**Comment by AdrK** [comment]
I have found that although ptrace and process_vm_* calls are not working, the direct read from /proc/<pid>/mem is working :)

https://github.com/AdrK/ptrace_test/blob/master/src/test_on_child.c

**Comment by AdrK** [comment]
At the end of the day it looks like Heroku is even more restricted as it also prohibits access to `/proc/<pid>/mem`. So omitting the seccomp security is not enough. We can make a phpspy work without `--cap-add` and `--security-opt seccomp=unconfined` and use direct read from /proc, but on Heroku it will not be enough.

**Comment by AdrK** [comment]
Quick check shows that direct read from /dev/<pid>/mem is possible on Heroku when the read it's done by parent process on itself, or child process.

**Comment by AdrK** [comment]
Here you can find a test application that works well on Heroku:
https://github.com/AdrK/ffi_test

It is a php script that is loading a phpspy as a shared object library and calls phpspy_init, phpspy_snapshot & phpspy_cleanup passing its own pid as an argument.

It was successfully deployed and ran on Heroku platform.

![obraz](https://user-images.githubusercontent.com/15175440/120810882-b97b0b80-c54b-11eb-9c0b-908943f4d57c.png)

**Comment by AdrK** [comment]
Speed comparision of process_vm_readv versus lseek & read. Looks like direct read is way slower.

![image](https://user-images.githubusercontent.com/15175440/121346574-932be600-c926-11eb-8d0a-6e9673fa83fa.png)
